Modem updates

kujaku11 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Update to functionality of Modem tools and updated documentation.  

## Description
<!--- Describe your changes in detail -->
- Added functions to Data
    - add/remove a station
    - remove component
    - add error to components
    - swap phase
    - remove static shift
    - write VTK file in right-hand or left-hand coordinate system
    - Added `ocean_bottom` argument in project_stations_on_topography to place stations at bottom of ocean 
- Added functions to Model
    - write files for Geosoft and Leapfrog
    - Added arguments to add_topography_to_model2 to shift east and north incase the projection is off (often when UTM grids are crossed)
- All station locations taken care of with Station
- added function `Station.write_shp_file` to write a shapefile of station locations
- Fixed bug in Station when stations cross UTM grids, will now project all stations to the center point UTM zone, can specify that zone or its calculated.
- Updated import of evtk to pyevtk
- Updated documentation 
- Updated tests  
- Added functionality to gui/modem_plot_response_qt5.py to include the added functionality in Data
- Updated tests to use Path

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Adding functionality
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. --> 
Added test for Data. 
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My code follows the code style of this project.
- [ x] My change requires a change to the documentation.
- [x ] I have updated the documentation accordingly.
- [x ] I have added tests to cover my changes.
- [x ] All new and existing tests passed 
- Well, Almost the rotation of mesh test fails, its only off by a few meters. So either recalculate or update default files.

kujaku11 committed:
copied over modem changes from metadata

kujaku11 committed:
cleaning up files to comply with develop

kujaku11 committed:
updated mtcolors to have terrain colormap

kujaku11 committed:
fixed bug in station finding center

kujaku11 committed:
updated the format of rotation angle to be .3g to comply with older modem file formats for testing

kujaku11 committed:
updated test_model and tests.__init__ to use Path

kujaku11 committed:
updated test_model and tests.__init__ to use Path

kujaku11 committed:
updated test_model to use path and EDI_DATA_LIST

kujaku11 committed:
update test_data to use Path and EDI_DATA_LIST

kujaku11 committed:
updated how center point elevation is estimate, should be a negative of the max elevation

kujaku11 committed:
fixed how center elevation is passed

kujaku11 committed:
updated mesh_tools with functions for rounding, and shifting a surface

kujaku11 committed:
updated documentation on mesh_tools.get_rounding

kujaku11 committed:
updated documentation in data file

kujaku11 committed:
updated documentation in modem.Data

kujaku11 committed:
added ability to write vtk stations in z+ or z- coordinate system

kujaku11 committed:
added check for ocean bottom sensor for issue #133

kujaku11 committed:
added check for ocean bottom sensor for issue #133

kujaku11 committed:
updated documentation and import of evtk to pyevtk

kujaku11 committed:
updated documentation and import of evtk to pyevtk

kujaku11 committed:
updated documentation of model

kujaku11 committed:
updated modem guis

kujaku11 committed:
updating test_data_object

kujaku11 committed:
updating gui/modem_*

alkirkby commented:
ModEM RMS maps seems to be broken on this branch now? I tried running the example /examples/scripts/ModEM_PlotRMS.py and it doesn't work on modem_updates but works on develop:

Traceback (most recent call last):

  File "C:\mtpywin\mtpy\examples\scripts\ModEM_PlotRMS.py", line 54, in <module>
    plot_elements='both',

  File "C:\mtpywin\mtpy\mtpy\modeling\modem\plot_rms_maps.py", line 240, in __init__
    self.plot()

  File "C:\mtpywin\mtpy\mtpy\modeling\modem\plot_rms_maps.py", line 427, in plot
    rms, filt = self._calculate_rms(p_dict)

  File "C:\mtpywin\mtpy\mtpy\modeling\modem\plot_rms_maps.py", line 259, in _calculate_rms
    :, self.period_index, ii, jj

IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

alkirkby commented:
I also get an error when trying to create a datafile: running example mtpy\examples\scripts\ModEM_build_inputfiles.py:

ERROR:mtpy.modeling.modem.data.Data:'utf-32-le' codec can't decode bytes in position 0-3: code point not in range(0x110000)
Traceback (most recent call last):

  File "C:\mtpywin\mtpy\mtpy\modeling\modem\data.py", line 895, in fill_data_array
    data_array = self.get_relative_station_locations(mt_dict, data_array)

  File "C:\mtpywin\mtpy\mtpy\modeling\modem\data.py", line 490, in get_relative_station_locations
    stations_obj.get_station_locations(mt_list)

  File "C:\mtpywin\mtpy\mtpy\modeling\modem\station.py", line 267, in get_station_locations
    self.calculate_rel_locations()

  File "C:\mtpywin\mtpy\mtpy\modeling\modem\station.py", line 284, in calculate_rel_locations
    self.station_locations["rel_east"] = self.east - self.center_point.east[0]

  File "C:\mtpywin\mtpy\mtpy\modeling\modem\station.py", line 359, in center_point
    self.logger.info(f"Projecting lat, lon to UTM zone {center_location['zone'][0]}")

UnicodeDecodeError: 'utf-32-le' codec can't decode bytes in position 0-3: code point not in range(0x110000)


During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File "C:\mtpywin\mtpy\examples\scripts\ModEM_build_inputfiles.py", line 78, in <module>
    do.write_data_file()

  File "C:\mtpywin\mtpy\mtpy\modeling\modem\data.py", line 1255, in write_data_file
    longitude_format=longitude_format,

  File "C:\mtpywin\mtpy\mtpy\modeling\modem\data.py", line 906, in fill_data_array
    raise ValueError(error)

ValueError: 'utf-32-le' codec can't decode bytes in position 0-3: code point not in range(0x110000)

kujaku11 committed:
updated modem_gui

kujaku11 commented:
@alkirkby RMS Plots fails because there is an option to set the attribute period_index to "all".  

From the logic it looks _calculate_rms only wants an integer to return the rms array.  So the question is what is meant by "all"?  Is it to plot each period separately, or is it supposed to be a single plot where all periods are averaged or something else?

If its plot a summarization of all periods in one plot, I added a fix using np.nanmean() and it works.

kujaku11 committed:
fixed plot_rms_maps assuming 'all' is meant to be a summarization of all periods into one plot

kujaku11 committed:
fixed plot_rms_maps assuming 'all' is meant to be a summarization of all periods into one plot

kujaku11 committed:
updated ModEM_build_inputsfiles to use Path and directory variables from \tests

kujaku11 commented:
Could not replicate your errors in ModEM_build_inputfiles, but I did update that example to import the directories for EDI files and where the output files go.  Also updated to plot the topography.

kujaku11 committed:
updated ModEM_build_inputfiles_edi_elevation.py to use Path and imported directories

kujaku11 committed:
updated C:\Users\jpeacock\Documents\GitHub\mtpy\examples\scripts\ModEM_build_inputfiles_rotate_grid.py to use Path and imported directories

kujaku11 committed:
update examples\scripts\ModEM_DepthSlice_to_gmt.py to use Path and imported directories and fixed but in modem.Model write_xyres to write the correct depth

kujaku11 committed:
update examples\scripts\ModEM_plot_slice_on_profile.py to use Path and imported directories

kujaku11 committed:
fixed test_mare2dem to skip comparing lines with more than one '-'

kujaku11 committed:
updated tests, they all pass now.

alkirkby commented:
> @alkirkby RMS Plots fails because there is an option to set the attribute period_index to "all".
> 
> From the logic it looks _calculate_rms only wants an integer to return the rms array. So the question is what is meant by "all"? Is it to plot each period separately, or is it supposed to be a single plot where all periods are averaged or something else?
> 
> If its plot a summarization of all periods in one plot, I added a fix using np.nanmean() and it works.

It's not a straight mean it's an RMS - I did some checking to make sure the overall RMS matches what is in the logfile in ModEM and it needs to be calculated as an RMS so i do the same for the by station RMS. There should be a bunch of properties in the rms_array in teh residual object. THe one you need for component is rms_array['rms_z_component']. Check out _calculate_rms, line 240 of rms_maps in develop

kujaku11 committed:
updated tests, they all pass now. Ran black

kujaku11 committed:
updated tests, they all pass on local machine

kujaku11 committed:
updated tests, they all pass on local machine

kujaku11 committed:
added response_plot_settings

kujaku11 committed:
updated plot_stations to have a nice looking station label

kujaku11 committed:
updated plotRMS and tests to follow alisons comment

kujaku11 committed:
updated plotRMS and tests to follow alisons comment

kujaku11 committed:
updated plotRMS and tests to follow alisons comment

kujaku11 commented:
> It's not a straight mean it's an RMS - I did some checking to make sure the overall RMS matches what is in the logfile in ModEM and it needs to be calculated as an RMS so i do the same for the by station RMS. There should be a bunch of properties in the rms_array in teh residual object. THe one you need for component is rms_array['rms_z_component']. Check out _calculate_rms, line 240 of rms_maps in develop

Sounds right, I updated accordingly.

kujaku11 committed:
updated occam2d to remove some of the warnings

kujaku11 committed:
updated mare2d tests and fixed some of the warnings for occam2d

kujaku11 committed:
trying a new commit to rerun travisCL

kujaku11 committed:
trying a new commit to rerun travixcl

alkirkby commented:
I am still having trouble getting modem_build_inputfiles.py to work. Same error as before. It seems to have a problem with center_location['zone']. Will keep looking into it

alkirkby commented:
@kujaku11  when you run the example code, what do you get for center_location['zone']? for me it is empty (None) and I think that's where the error is coming in but I'm not completely sure. It's frustrating as sometimes it gives an error when I try and print center_location['zone'] so really hard to debug

kujaku11 commented:
> @kujaku11 when you run the example code, what do you get for center_location['zone']? for me it is empty (None) and I think that's where the error is coming in but I'm not completely sure. It's frustrating as sometimes it gives an error when I try and print center_location['zone'] so really hard to debug

@alkirkby  That is odd.  When I run  `mtpy\examples\scripts\ModEM_build_inputfiles.py` I get the following:

```
ModEM Data Object:
	Number of stations: 28
	Number of periods:  26
	Period range:  
		Min: 0.0017782794100389228 s
		Max: 3162.2776601683795 s
	Rotation angle:     0.0
	Data center:        
		 latitude:  -20.5100 deg
		 longitude: 138.0100 deg
		 Elevation: -516.0 m
		 Easting:   188142.4792 m
		 Northing:  7729227.6114 m
		 UTM zone:  54K
	Model EPSG:         28354
	Model UTM zone:     None
	Impedance data:     True
	Tipper data:        True
```

And get the following plots:

![modem_build_inputfiles_mesh_01](https://user-images.githubusercontent.com/2100870/108805139-2b5ab680-7554-11eb-9445-ce8fa1eff6fc.png)
![modem_build_inputfiles_topo_01](https://user-images.githubusercontent.com/2100870/108805141-2b5ab680-7554-11eb-9be9-b9bc8e79a61f.png)


Have you pulled the current version?

alkirkby commented:
Yep I pull the latest version each day, if you print center_location['zone'] what do you get? is it None?

kujaku11 commented:
> Yep I pull the latest version each day, if you print center_location['zone'] what do you get? is it None?

```
do.center_point
Out[9]: 
rec.array([ (-20.51,  138.01,  188142.47916241,  7729227.61144812, -516., '54K')], 
          dtype=[('lat', '<f8'), ('lon', '<f8'), ('east', '<f8'), ('north', '<f8'), ('elev', '<f8'), ('zone', '<U4')])
```

```
do.station_locations.center_point
Out[10]: 
rec.array([ (-20.51,  138.01,  188142.47916241,  7729227.61144812, -516., '54K')], 
          dtype=[('lat', '<f8'), ('lon', '<f8'), ('east', '<f8'), ('north', '<f8'), ('elev', '<f8'), ('zone', '<U4')])
```

kujaku11 commented:
And the error you are getting is:

ValueError: 'utf-32-le' codec can't decode bytes in position 0-3: code point not in range(0x110000)

That sounds like an encoding issue.  What type of machine are you on?

kujaku11 commented:
What do you get if you `print do.station_locations.utm_zone`?

alkirkby commented:
the do.station_locations object is None. Perhaps we should catch up on skype to try and sort this out?

kujaku11 committed:
added print statements in station to help debug

kujaku11 committed:
added print statements in station to help debug

kujaku11 committed:
added print statements in station to help debug

kujaku11 committed:
updated station to explicitly set center_point[0]

kujaku11 committed:
Merge branch 'develop' into modem_updates

kujaku11 commented:
I've finished most of the updates to the ModEM module.  Merged with the most recent version of develop.

kujaku11 committed:
updated calculator to be same as develop branch

kujaku11 commented:
There was a merge conflict that I messed up in utils/calculator.py with rounding numbers.  Updated to what is on the develop branch.

kujaku11 committed:
added __str__ and __repr__ to modem.Model

alkirkby commented:
just running the example ModEM_plot_slices_on_basemap.py and it produces nothing in the plot, e.g.: 
![DepthSlice2km](https://user-images.githubusercontent.com/5335745/109732415-3dea8500-7c11-11eb-8ca9-43255f1ede31.png)

Compare with output from develop branch: 
![DepthSlice2km](https://user-images.githubusercontent.com/5335745/109732490-64102500-7c11-11eb-8d5d-216d2848adaf.png)

I will continue testing and look into it at the end.

alkirkby commented:
examples/scripts/ModEM_PlotPTmap.py needed a small tweak (on this branch and develop) so that files save correctly. I have made the change on the develop branch.

alkirkby commented:
Everything else works fine for me so once basemap plot is resolved should be good to go. Sorry I introduced conflicts - I have updated occam2d to fix issue #135 so I think that just needs to be brought across but check you are happy with it.

kujaku11 committed:
added to_csv, to_shp for modem.Station

kujaku11 committed:
updated examples/scripts/ModEM_plot_slices_on_basemap.py to use relative paths, haven't found the issue yet with the plotting

kujaku11 committed:
added an MT jet like color scale

kujaku11 commented:
> Everything else works fine for me so once basemap plot is resolved should be good to go. Sorry I introduced conflicts - I have updated occam2d to fix issue #135 so I think that just needs to be brought across but check you are happy with it.

Any resolution on this, I'm having issues installing mpl_toolkits.basemap and haven't been able to test it.

kujaku11 committed:
Merge branch 'develop' into modem_updates

alkirkby commented:
Ah sorry no I haven't looked into it. You need to install basemap e.g. conda install basemap (I think) then you should be able to get the modules from mpl_toolkits. I will look at it now

alkirkby commented:
I found the bug but need to go now unfortunately so haven't managed to fix it. It's plotting in the wrong spot because the center point is projected from lat/long to east/north using the utm zone of that center point, not the model_epsg that is optionally passed into the data object. It should just be a minor tweak (hopefully)

kujaku11 committed:
added function to plot conductance as a raster in array2raster

kujaku11 committed:
added ability to plot conductance to the shape files

kujaku11 committed:
Merge branch 'modem_updates' of https://github.com/MTgeophysics/mtpy into modem_updates
